crates/factor-outbound-networking/src/lib.rs

```diff
@@ -0,0 +1,116 @@
+            .map(|component| {
+                Ok((
+                    component.id().to_string(),
+                    component
+                        .get_metadata(ALLOWED_HOSTS_KEY)?
+                        .unwrap_or_default()
+                        .into_boxed_slice()
+                        .into(),
+                ))
+            })
+            .collect::<Result<_>>()?;
+        Ok(AppConfig {
+            component_allowed_hosts,
+        })
+    }
+}
+
+#[derive(Default)]
+pub struct AppConfig {
+    component_allowed_hosts: HashMap<String, Arc<[String]>>,
+}
+
+type AllowedHostsFuture = Shared<BoxFuture<'static, Arc<anyhow::Result<AllowedHostsConfig>>>>;
+
+pub struct InstancePreparer {
+    allowed_hosts_future: AllowedHostsFuture,
+}
+
+impl FactorInstancePreparer<OutboundNetworkingFactor> for InstancePreparer {
+    fn new<Factors: SpinFactors>(
+        ctx: PrepareContext<OutboundNetworkingFactor>,
+        mut preparers: InstancePreparers<Factors>,
+    ) -> Result<Self> {
+        let hosts = ctx
+            .app_config()
+            .component_allowed_hosts
+            .get(ctx.app_component().id())
+            .cloned()
+            .context("missing component allowed hosts")?;
+        let resolver = preparers.get_mut::<VariablesFactor>()?.resolver().clone();
+        let allowed_hosts_future = async move {
+            let prepared = resolver.prepare().await?;
+            AllowedHostsConfig::parse(&hosts, &prepared)
+        }
+        .map(Arc::new)
+        .boxed()
+        .shared();
+        // let prepared_resolver = resolver.prepare().await?;
+        // let allowed_hosts = AllowedHostsConfig::parse(
+        //         .context("missing component allowed hosts")?,
+        //     &prepared_resolver,
+        // )?;
+
+        // Update Wasi socket allowed ports
+        let wasi_preparer = preparers.get_mut::<WasiFactor>()?;
+        let hosts_future = allowed_hosts_future.clone();
+        wasi_preparer.outbound_socket_addr_check(move |addr| {
+            let hosts_future = hosts_future.clone();
+            async move {
+                match &*hosts_future.await {
+                    Ok(allowed_hosts) => {
+                        // TODO: verify this actually works...
+                        spin_outbound_networking::check_url(&addr.to_string(), "*", allowed_hosts)
+                    }
+                    Err(err) => {
+                        // TODO: should this trap (somehow)?
+                        tracing::error!(%err, "allowed_outbound_hosts variable resolution failed");
+                        false
+                    }
+                }
+            }
+        });
+        Ok(Self {
+            allowed_hosts_future,
+        })
+    }
+
+    fn prepare(self) -> Result<<OutboundNetworkingFactor as Factor>::InstanceState> {
+        Ok(())
+    }
+}
+
+impl InstancePreparer {
+    pub async fn resolve_allowed_hosts(&self) -> Arc<anyhow::Result<AllowedHostsConfig>> {
```

Comment:
I think this is no longer used?

Reply:
This will be used by the e.g. outbound-http factor.